sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,365 @@
+in a `sub_group`.  The `invoke_simd` interface marshals data between the
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+template <typename Callable, typename...T>
+auto invoke_simd(sub_group sg, Callable&& f, T... args);
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+The `Callable` passed to `invoke_simd` must be a C++ callable that can be
+invoked with the same number of arguments specified in the `args` parameter
+pack. `Callable` may be a function object, a lambda,
+or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
+`Callable` must be an immutable callable with the same type and state for all
+work-items in the sub-group, otherwise behavior is undefined.
+
+Each argument to the invoked `Callable` must be a scalar type, a SIMD type,
+or a `std::tuple` of these types. The length of each SIMD type must match the
+sub-group size of the kernel; calling a SIMD function with mismatched lengths
+is an error, and an implementation must issue a diagnostic.
+
+To enable overloading in cases where the `Callable` accepts either a scalar
+type or a SIMD type, we leverage the `sycl::ext::oneapi::uniform` wrapper
+type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.
+
+Each argument in the `args` parameter pack must be an arithmetic type, a
+trivially copyable type wrapped in a `sycl::ext::oneapi::uniform`, or a
+`std::tuple` of these types.  Arguments may not be pointers or references,
+but pointers (like any other non-arithmetic type) may be passed if wrapped in a
+`sycl::ext::oneapi::uniform`.  Any such pointer value must point to memory
+that is accessible by all work-items in the sub-group (i.e. the pointer must
+point to an allocation in local or global memory).  The address space for such
+pointers can be local, global or generic; if a generic pointer that points
+to an allocation in private memory is passed as an argument, the behavior is
+undefined.
+
+In order to invoke the SIMD function, `invoke_simd` converts each argument
+in the `args` parameter pack according to the following rules:
+
+- Arguments of type `bool` are converted to type
+  `sycl::ext::oneapi::simd_mask<bool, N>`, where `N` is the sub-group size of
+  the calling kernel.  Element `i` of the SIMD type represents the value from
+  the work-item with sub-group local ID `i`.
+
+- All other arithmetic arguments of type `T` are converted to type
+  `sycl::ext::oneapi::simd<T, N>`, where `N` is the sub-group size of the
+  calling kernel.  Element `i` of the SIMD type represents the value from the
+  work-item with sub-group local ID `i`.
+
+- Arguments of type `sycl::ext::oneapi::uniform<T>` are converted to type `T`.
+  The value may be taken from any work-item in the sub-group; wrapping an
```

Comment:
This needs to be clarified for cases when the call is under divergent condition. Calling the function under divergent condition means that not all work items are valid.

Reply:
Good point.

I think we should update this to say that a `uniform` can only be constructed in converged control flow (at least for now).  All of the use-cases we've seen for this tie into sub-group functions that already have this requirement.

Reply:
Fixed in d47e0094de20c26dfb797401a9819cb8a9dfa83b.